### PR TITLE
Start uncythonization

### PR DESCRIPTION
**Pull request by martindurant**
Notes:
- most underscored attributes with property attributes have been replaced by simple non-underscored attributes, since they were non-uniformly accessed using both modes; a few need actions on set, so those have been maintained. This change may even improve performance slightly for fewer name lookups.
- attributes of the Scheduler and underscored methods have *not* been changed
- the class hierarchy has not been reverted yet, and I am not convinced we need to
- we have lost some type information, which was mostly in the form of cython types; mypy complains a lot (but how did it ever pass??). Existing types are inconsistent in a number of places, since `__init__`s set a `None` value and this is immediately overwritten at the site of instance creation, e.g., `.prefix` in `new_task()`. 

- [x] Closes https://github.com/dask/community/issues/228 , https://github.com/dask/distributed/pull/5839 , we had a community issue about this?
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/scheduler.py
Comment (mrocklin) [inline distributed/scheduler.py:5099]: Woot

File: distributed/scheduler.py
Comment (mrocklin) [inline distributed/scheduler.py:1145]: Losing these properties makes me very happy 😄

**Review by mrocklin — COMMENTED** [review]
What remains to be done after this?

**Comment by martindurant** [comment]
@mrocklin , added update to the description. Currently all tests in test_scheduler.py pass bar one. Of course, it would be nice to push this through quickly, else it'll cause major merge conflicts.

**Comment by martindurant** [comment]
(now all tests in test_scheduler pass; but there are more tests in general which I haven't even tried yet, and of course appeasing mypy)

**Comment by mrocklin** [comment]
Thank you for your effort here.  I agree with the need to squeeze this in.

**Comment by jakirkham** [comment]
Thanks for working on this Martin!

**Comment by jakirkham** [comment]
Maybe we should close these issues as well?

* https://github.com/dask/community/issues/228
* https://github.com/dask/distributed/issues/4442
* https://github.com/dask/distributed/issues/5685
* https://github.com/dask/distributed/pull/4418
* https://github.com/dask/distributed/pull/4650
* https://github.com/dask/distributed/pull/4767
* https://github.com/dask/distributed/pull/4771

Also maybe this one?

* https://github.com/dask/distributed/issues/854

**Comment by jakirkham** [comment]
Also would be good to revert PR ( https://github.com/dask/distributed/pull/5831 )

**Comment by martindurant** [comment]
> Also would be good to revert PR ( https://github.com/dask/distributed/pull/5831 )

bahh yeah, I hope there's no too much of that

**Comment by jakirkham** [comment]
Shouldn't be, but please let me know if you run into issues.

Would be tempted to just try `git revert 94ebd5711c31a709aa1bb42ba3b6d8dfbd3c63ad`.

File: .github/workflows/conda.yml
Comment (jakirkham) [inline .github/workflows/conda.yml:11]: Would hang onto these

**Comment by jakirkham** [comment]
@charlesbluca, could you please look at the Conda package changes here to make sure we haven't missed anything?

File: .github/workflows/conda.yml
Comment (martindurant) [reply .github/workflows/conda.yml:11]: They seem to be back https://github.com/dask/distributed/pull/6104/files#diff-72b980456954eec0c566d17ce7746f2c1f7c6b9b98b1ad01395b6f7cb8c4ca0b (I think the merge was a bit weird here...)

Comment (jakirkham) [reply .github/workflows/conda.yml]: Ah ok. Yep looks better now. Thanks Martin!

File: .github/workflows/conda.yml
Comment (charlesbluca) [inline .github/workflows/conda.yml:81]: Ideally we should keep distributed nightlies `noarch`, so that we don't need to convert the package to different architectures before uploading:

```suggestion
```

File: .github/workflows/conda.yml
Comment (charlesbluca) [inline .github/workflows/conda.yml:91]: Wouldn't need to upload Linux packages anymore since that was only for the Cython builds:

```suggestion
```

File: .github/workflows/conda.yml
Comment (charlesbluca) [inline .github/workflows/conda.yml:66]: With the above changes, we would no longer need to build in a specific directory:

```suggestion
                           --output-folder .
```

File: .github/workflows/conda.yml
Comment (charlesbluca) [inline .github/workflows/conda.yml:60]:
```suggestion
                           --output-folder .
```

File: continuous_integration/recipes/distributed/meta.yaml
Comment (charlesbluca) [inline continuous_integration/recipes/distributed/meta.yaml:24]: We no longer need `distributed-impl` now that we aren't doing cython builds correct?

If so, we should probably replace everything below `source` with the contents of the `distributed` output

**Comment by martindurant** [comment]
Some green! :)

**Comment by martindurant** [comment]
Would anyone here like to help me mypy this out? See the last bullet of the description.

File: continuous_integration/recipes/distributed/meta.yaml
Comment (martindurant) [reply continuous_integration/recipes/distributed/meta.yaml:24]: I fixed the other points, above, but I didn't understand what to do here.

Comment (charlesbluca) [reply continuous_integration/recipes/distributed/meta.yaml:24]: Yeah I realize I probably didn't explain this the best 😅 opened a PR against your branch doing what I described:

- https://github.com/martindurant/distributed/pull/1

**Comment by martindurant** [comment]
OK, just types left. Who votes for liberal "ignore" to move forward versus those for sprinkling types around?

**Comment by jakirkham** [comment]
Personally am ok with that approach 👍

**Comment by mrocklin** [comment]
> OK, just types left. Who votes for liberal "ignore" to move forward versus those for sprinkling types around?

Can you say more about what you mean about having "types left"?  Is this mypy complaining to you, or are there still cython types lying around?

Also, just to clarify on scope, full decythonization (which I don't think you're on the hook for, to be clear) would also involve rejoining things like SchedulerState, decide_worker, and so on. There were a variety of code structure changes that were done to accomdate Cython which we would probably want to undo.  (although again, you're not on the hook for this solo)

**Comment by martindurant** [comment]
I meant mypy - the reason that the lint CI run shows red. I have fixed some locally, we'll see how far I get tonight. I'm not sure why they weren't picked up before.

> to clarify on scope

Yes, agreed, all the changes I want to make in this PR are ready except mypy types. Actually, I thought it could do with some thinking about whether we want to remerge the classes or not - to be had in a separate discussion.

**Comment by martindurant** [comment]
Down to 16 missing type annotations and no errors

File: distributed/scheduler.py
Comment (fjetter) [inline distributed/scheduler.py:556]: we might still want to use slots?

**Comment by mrocklin** [comment]
I'm fine with liberal use of `#type: ignore` if we think that it's hiding nitpicky rather than structural issues.  We can think about unleashing @crusaderky on this part of the codebase after things get structurally cleaned up.

**Review by crusaderky — CHANGES_REQUESTED** [review]

**Comment by martindurant** [comment]
General comments
- on class-level annotations: all of these that were deleted were cython types, and not sufficient for mypy to be able to validate agains. Hence, they are gone, but I suggest a separate follow-up PR could be used to add them, if someone has the spare capacity.
- on `att : str = None # type: ignore` constructs; these are bad and should not be written this way. They are *not* `str | None` types in most cases, since instantiating code immediately sets the value after `__init__` returns. This suggests we need a rewrite, but I'd rather not do that here.

File: distributed/scheduler.py
Comment (martindurant) [reply distributed/scheduler.py:1751]: why remove these?

**Comment by jakirkham** [comment]
Right got the sense that the reason Coiled requested dropping Cythonization is that there is other work Coiled wants to pursue in the Scheduler. Assuming I've understood this correctly (which I may not, please feel free to correct me), it seems like we should decide on a handoff point. It seems like Martin has done a great job at dropping Cython already. So think we are very close to (if not already at) this handoff point.

File: distributed/scheduler.py
Comment (crusaderky) [reply distributed/scheduler.py:1751]: Redundant. When you create a new variable its type is automatically inferred from whatever it's assigned to it. In this case a bool. Conceptually it's the same as ``auto`` in C++11.

File: distributed/scheduler.py
Comment (martindurant) [reply distributed/scheduler.py:165]: If it's not in `__slots__,` you can't access it at all, no?

File: distributed/scheduler.py
Comment (crusaderky) [reply distributed/scheduler.py:1751]: There are many more that could be cleaned up later - e.g.
``ts: TaskState = self.tasks[key]`` could be simplified, but only after upgrading the annotation of ``tasks`` from ``dict`` to ``dict[str, TaskState]``. It's labour that I'm happy to deal with in a successive PR.

**Comment by crusaderky** [comment]
> * on class-level annotations: all of these that were deleted were cython types, and not sufficient for mypy to be able to validate agains. Hence, they are gone, but I suggest a separate follow-up PR could be used to add them, if someone has the spare capacity.

OK, I'll reinstate them in a later PR.

> * on `att : str = None # type: ignore` constructs; these are bad and should not be written this way. They are _not_ `str | None` types in most cases, since instantiating code immediately sets the value after `__init__` returns. This suggests we need a rewrite, but I'd rather not do that here.

I agree they should be dealt with at a later stage.

**Comment by mrocklin** [comment]
> Right got the sense that the reason Coiled requested dropping Cythonization is that there is other work Coiled wants to pursue in the Scheduler

Not quite.  We've identified that the main bottleneck to use today isn't performance, but rather stability.  We're not really doing anything new these days.  We're mostly trying to keep the lights on (fight deadlocks, get CI working again, figure out how to keep memory under control, and so on).  There was significant complexity accrued in the Cythonization effort without really providing a ton of value for a while, and so we're arguing to drop it.  That's ok, not every experiment needs to succeed.

Mostly I want to avoid the impression that there is some secret work happening somewhere for which this is required.  We're in full maintenance mode right now.  Help on that maintenance is welcome, especially from folks who know a lot about what happened in this experiment.  We can take all of of the maintenance though if other folks are too busy.

File: distributed/scheduler.py
Comment (jakirkham) [inline distributed/scheduler.py:115]: This function was a fallback for when Cython wasn't available . We can drop it.

```suggestion
```

**Comment by jakirkham** [comment]
> Mostly I want to avoid the impression that there is some secret work happening somewhere for which this is required.

No worries. Did not have that impression. Florian had mentioned in one of our calls doing similar cleanup with the Scheduler as is being done with the Worker. So this is what I was thinking of here. Don't really know what that would entail so was sparse in details in my previous comment.

File: distributed/scheduler.py
Comment (martindurant) [reply distributed/scheduler.py:1739]: OK, marking this as resolved then. Thanks.

File: distributed/scheduler.py
Comment (martindurant) [reply distributed/scheduler.py:32]: 👍

**Comment by martindurant** [comment]
All good, @crusaderky ?

File: distributed/scheduler.py
Comment (crusaderky) [reply distributed/scheduler.py:153]: if it's not in ``__slots__`` it cannot be _stored_ at all, privately or not.
What I'm suggesting is that ``_hash`` was marked before as something that other classes shouldn't access, and it should remain that way.

Comment (martindurant) [reply distributed/scheduler.py]: understood - done.

File: distributed/scheduler.py
Comment (martindurant) [reply distributed/scheduler.py]: oops

Comment (crusaderky) [reply distributed/scheduler.py]: I think this is the last bit before it can be merged :)

**Review by crusaderky — APPROVED** [review]
Ready to merge as soon as CI is green

**Comment by crusaderky** [comment]
Thank you

**Pull request merged**
